pkg/resource/stack/deployment.go

```diff
@@ -671,14 +647,21 @@ func DeserializePropertyValue(v interface{}, dec config.Decrypter,
 					if err := json.Unmarshal([]byte(plaintext), &elem); err != nil {
 						return resource.PropertyValue{}, err
 					}
-					ev, err := DeserializePropertyValue(elem, config.NopDecrypter, enc)
+					ev, err := DeserializePropertyValue(elem, config.NopDecrypter)
 					if err != nil {
 						return resource.PropertyValue{}, err
 					}
 					prop := resource.MakeSecret(ev)
 					// If the decrypter is a cachingCrypter, insert the plain- and ciphertext into the cache with the
 					// new *resource.Secret as the key.
-					if cachingCrypter, ok := dec.(*cachingCrypter); ok {
+					if cachingCrypter, ok := dec.(*cachingSecretsManager); ok {
+						if !cipherOk {
+							encryptedText, err := cachingCrypter.EncryptValue(ctx, plaintext)
```

Comment:
Curiously, I'm not 100% clear on exactly how we currently reach this line because the `DeserializeDeploymentV3` method always wraps the type in a `mapDecrypter` which wouldn't have matched the old type check (nor the new one either). Overall, I think we'll also want to fold the `mapDecrypter` into the `cachingSecretsManager`.